core/src/main/java/hudson/model/ParametersAction.java

```diff
@@ -302,15 +323,37 @@ public void onLoad(Run<?, ?> r) {
     }
 
     private boolean isSafeParameter(String name) {
+        loadSafeParameters();
+        return safeParameters.contains(name);
+    }
+
+    /**
+     * Combines the contents of {@link #SAFE_PARAMETERS_SYSTEM_PROPERTY_NAME}
+     * and {@link #getAdditionalSafeParameters()} into {@link #safeParameters}.
+     * @since TODO
+     */
+    private void loadSafeParameters() {
         if (safeParameters == null) {
             String paramNames = SystemProperties.getString(SAFE_PARAMETERS_SYSTEM_PROPERTY_NAME);
+            safeParameters = new TreeSet<>();
             if (paramNames != null) {
-                safeParameters = Arrays.asList(paramNames.split(","));
-            } else {
-                safeParameters = Collections.emptyList();
+                safeParameters.addAll(Arrays.asList(paramNames.split(",")));
             }
+            safeParameters.addAll(getAdditionalSafeParameters());
         }
-        return safeParameters.contains(name);
+    }
+
+    /**
+     * Provides a list of parameter names considered safe by the class overriding this action.
+     * Plugins can extend this when scheduling a build with the built in parameters it has.
+     * Whatever the user provides in {@link #SAFE_PARAMETERS_SYSTEM_PROPERTY_NAME} or
+     * {@link #KEEP_UNDEFINED_PARAMETERS_SYSTEM_PROPERTY_NAME} still counts.
+     *
+     * @return an additional list of safe parameter names
+     * @since TODO
+     */
+    protected Collection<String> getAdditionalSafeParameters() {
+        return Collections.emptyList();
```

Comment:
Well, in that case you wouldn't need to extend the original action which is a plus. But would that make it too simple and potentially open a new hole for undisciplined maintainers?

It would make it slightly harder, for those like me, who wants to stay on an older core dependency but have the safe parameters activated when running on a newer core since you'd need to use some reflection.

But I like the idea better than mine, just some concerns.